packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -606,8 +607,18 @@ mixin LocalHistoryRoute<T> on Route<T> {
     _localHistory.remove(entry);
     entry._owner = null;
     entry._notifyRemoved();
-    if (_localHistory.isEmpty)
-      changedInternalState();
+    if (_localHistory.isEmpty) {
+      if (SchedulerBinding.instance.schedulerPhase == SchedulerPhase.persistentCallbacks) {
+        // The local history might be removed as a result of disposing inactive
+        // elements during finalizeTree. The state is locked at this moment, and
+        // we can only notify state has changed in the next frame.
+        SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
+          changedInternalState();
```

Comment:
What if the route has been disposed at this point?

Reply:
it will be fine surprisingly, the https://github.com/flutter/flutter/blob/60b9d154957fdbc49ab0bea1937b3049ebe91bc9/packages/flutter/lib/src/widgets/routes.dart#L866 will prevent a setstate if the route is disposed -> ModalScope has been disposed. However, it seems dangerous to call changedInternalState on a disposed route. The Route also does not have a flag to indicate if it has been disposed. I am not sure what makes more sense here. Leave this as is, or expose something from Route like `debugIsDisposed`.